File: templates/backend/fastapi-starter/src/core/email.py
"""Secure HTTPS delivery for credentials that must never enter API responses."""
import asyncio
import json
from urllib import error, request

from src.core.config import settings


class EmailDeliveryError(RuntimeError):
    pass


async def send_secure_email(*, recipient: str, subject: str, body: str) -> None:
    """Send through the shared authenticated adapter. Development may suppress mail."""
    if not settings.email_delivery_configured:
        if settings.environment in {"production", "staging"}:
            raise EmailDeliveryError("Secure email delivery is not configured.")
        return

    def deliver() -> None:
        payload = json.dumps({"to": recipient, "subject": subject, "text": body}).encode("utf-8")
        email_request = request.Request(
            settings.email_delivery_url,
            data=payload,
            headers={
                "Authorization": f"Bearer {settings.email_delivery_token}",
                "Content-Type": "application/json",
            },
            method="POST",
        )
        with request.urlopen(email_request, timeout=10) as response:
            if not 200 <= response.status < 300:
                raise EmailDeliveryError("Secure email delivery is unavailable.")

    try:
        await asyncio.to_thread(deliver)
    except EmailDeliveryError:
        raise
    except (error.URLError, OSError, TimeoutError, ValueError) as exc:
        raise EmailDeliveryError("Secure email delivery is unavailable.") from exc

File: templates/backend/fastify-starter/src/email.ts
import type { Config } from './config.js';

export class EmailDeliveryError extends Error {
  constructor() {
    super('Secure email delivery is unavailable.');
  }
}

export interface EmailSender {
  send(recipient: string, subject: string, text: string): Promise<void>;
}

/** Sends mail through an authenticated HTTPS transactional-email adapter. */
export class SecureEmailSender implements EmailSender {
  constructor(private readonly config: Config) {}

  async send(recipient: string, subject: string, text: string): Promise<void> {
    const { EMAIL_DELIVERY_URL: url, EMAIL_DELIVERY_TOKEN: token } = this.config;
    if (!url || !token) {
      if (this.config.NODE_ENV === 'production') throw new EmailDeliveryError();
      return;
    }
    try {
      const response = await fetch(url, {
        method: 'POST',
        headers: { Authorization: `Bearer ${token}`, 'Content-Type': 'application/json' },
        body: JSON.stringify({ to: recipient, subject, text }),
        signal: AbortSignal.timeout(10_000),
      });
      if (!response.ok) throw new EmailDeliveryError();
    } catch (error) {
      if (error instanceof EmailDeliveryError) throw error;
      throw new EmailDeliveryError();
    }
  }
}
